config/Config-kernel.in

```diff
@@ -1418,8 +1418,7 @@ config KERNEL_JFFS2_FS_SECURITY
 
 config KERNEL_WERROR
 	bool "Compile the kernel with warnings as errors"
-	default BUILDBOT
```

Comment:
Its additional QA check on buildbots, if for whatever reason the CI on GH doesn't spot the issue.  There are still some https://github.com/openwrt/actions-shared-workflows/pull/5

Reply:
I'm sorry, but if supposed to answer my question https://github.com/ynezz/openwrt/commit/b943f9141fa87abc8529d9f85307357c14bf7a4b#r140612056, then it doesn't answer it.

I still don't understand the motivation behind having `KERNEL_WERROR` enabled on buildbots as well as on the *current* GCC version (which we will always need to track now also in `config/Config-kernel.in`, as you explain in https://github.com/ynezz/openwrt/commit/b943f9141fa87abc8529d9f85307357c14bf7a4b#diff-84d9df6894b269940af0e3cea01b35970a0967e5a856c9666f147b1379c10771R5) but not just always and by default. My understand is that the choice to *always* enable `KERNEL_WERROR` on the buildbots just makes it *harder* and *less likely* to spot such errors *before* breaking kmod package builds. Not enabling it for future or outdated GCC versions makes some limited sense to me, but as buildbots will anyway always chose the currently selected-by-default version of GCC, what is left are local builds of developers or downstream projects which may always chose to disable `KERNEL_WERROR` in their local `.config` anway.

Hence it should either be enabled by default or disabled by default, I don't see the additional value added justifying the extra work of having to track the current gcc version in `config/Config-kernel.in`, nor do I see a reason for the buildbot-specific default to diverge from the non-buildbot default.

I appreciate all those CI improvements in general, but please, either explain this choice in more detail or lets go with either
a. Set `default y` and force everyone to deal with the fallout (not bad imho)
or
b. If we decide that's too much work, lets have it optional and not set `default y`

Reply:
> as well as on the current GCC version (which we will always need to track now also in config/Config-kernel.in

This is a trade-off between two extremes:

1. Having it fully enabled - bad user experience
2. Having it full disabled - bad developer experience, lower QA feedback etc.

In other words, this is opt-out for currently supported default GCC version, seems to work so far. 

> My understanding is that the choice to always enable KERNEL_WERROR on the buildbots just makes it harder and less likely to spot such errors before breaking kmod package builds. 

Actually, I still don't have a [complete picture](https://github.com/openwrt/openwrt/pull/15064#issuecomment-2037166170) about that [ltq-vmmc issue](https://github.com/openwrt/openwrt/pull/15065) you've fixed. The build log you've provided was from 23.05 buildbot and I don't see any compiler issue there, but maybe I'm blind, so can you perhaps help me and tell me which line exactly shows that error?

That [package build step](https://buildbot.openwrt.org/images/#/builders/22/builds/54/steps/34/logs/stdio) you pointed me to was passing (`18:57 Packages built`), so I'm wondering how is that even possible?

Is the Makefile for the `ltq-vmmc` package broken? I mean, once the build of package fails, the whole package build step should fail and thus the whole target build would fail and IRC bot would notify #openwrt-devel about that breakage.

> what is left are local builds of developers or downstream projects which may always chose to disable KERNEL_WERROR in their local .config anyway.

Yes, that's the option, but ideally it should work out of the box. If it doesn't, lets hope there is a bug report filed and the problem fixed. 

> Set default y and force everyone to deal with the fallout (not bad imho)

The past clearly shows us, that this is a bad idea.

> If we decide that's too much work, lets have it optional and not set

IMO the current trade-off somehow works. AFAIK we don't have a lot of such compile bug reports from users anymore, but developers are still able to spot the issues and fix them.

Does it really matter to you, where we enable this option for buildbots? IMO its better to have this explicit `default BUILDBOT` in the tree, than following in `ansible-build.git` project (which is not public): 

```diff
diff --git a/inventories/prod/group_vars/all/02-openwrt-prod.yml b/inventories/prod/group_vars/all/02-openwrt-prod.yml
index 5b3a398f5c21..c652de6031dc 100644
--- a/inventories/prod/group_vars/all/02-openwrt-prod.yml
+++ b/inventories/prod/group_vars/all/02-openwrt-prod.yml
@@ -23,6 +23,7 @@ buildmaster:
     seedconfig: |-
       CONFIG_BUILDBOT=y
       CONFIG_DEVEL=y
+      CONFIG_KERNEL_WERROR=y
       # CONFIG_CCACHE is not set
       CONFIG_KERNEL_KALLSYMS=y
       CONFIG_AUTOREMOVE=y
```

I would still propose to keep it enabled on the buildbot, the feedback is wanted.

Reply:
First of all, I agree that it's good to configure buildbots transparently via `CONFIG_BUILDBOT`. Doing the same in a non-public YaML file somewhere hidden would be much much worse, for sure. However, I never argued for hiding the fact that this option is enabled on buildbots in non-public files and that was not what the whole discussion was all about, supposedly, at least from my point of view.

The thing is simple: Not having it enabled by default while having it enabled in the buildbots results in having things break on the buildbots which do not break in a local developer checkout (by default). I believe this divergence of default doesn't do a favor to anyone. 

To illustrate, let me tell how I ended up looking into this:
I discovered there is a problem with ltq-vmmc because @aparcar was contacting me and asking *"why can't I install kmod-ltql-vmmc? It seems missing in snapshot builds, but when I build everything locally it just works."*. And I was like *"let me check"*. I immediately suspected the switch to gcc-13 and thought that maybe @aparcar might still be using the gcc-12 toolchain because he may not have wiped `staging_dir` since the gcc update. So I've also tried building it locally, and it worked (because `CONFIG_KERNEL_WERROR` wasn't enabled). I then went on and downloaded [26 megabytes (!) of package build log](https://buildbot.openwrt.org/images/api/v2/logs/1168639/raw) and searched for the string `ltq-vmmc`. And there it was: the package was not built because warnings were treated as errors on the buildbots **but not by my local build** by default.
```
/builder/shared-workdir/build/staging_dir/host/bin/gcc -pthread -fPIC -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -O2 -I/builder/shared-workdir/build/staging_dir/host/include -I/builder/shared-workdir/build/staging_dir/hostpkg/include -I/builder/shared-workdir/build/staging_dir/target-mips_24kc_musl/host/include -I/builder/shared-workdir/build/staging_dir/host/include/e2fsprogs -O2 -I/builder/shared-workdir/build/staging_dir/host/include -I/builder/shared-workdir/build/staging_dir/hostpkg/include -I/builder/shared-workdir/build/staging_dir/target-mips_24kc_musl/host/include -I/builder/shared-workdir/build/staging_dir/host/include/e2fsprogs -fno-semantic-interposition -std=c11 -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declaration -fvisibility=hidden -fprofile-generate -I./Include/internal -I./Include -I. -I/builder/shared-workdir/build/staging_dir/host/include -I/builder/shared-workdir/build/staging_dir/hostpkg/include -I/builder/shared-workdir/build/staging_dir/host/include/e2fsprogs -I/builder/shared-workdir/build/build_dir/hostpkg/Python-3.11.7/Include -I/builder/shared-workdir/build/build_dir/hostpkg/Python-3.11.7 -c /builder/shared-workdir/build/build_dir/hostpkg/Python-3.11.7/Modules/unicodedata.c -o build/temp.linux-x86_64-3.11/builder/shared-workdir/build/build_dir/hostpkg/Python-3.11.7/Modules/unicodedata.o
/builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0/src/mps/drv_mps_vmmc_common.c:392:14: error: conflicting types for 'ifx_mps_fastbuf_init' due to enum/integer mismatch; have 'IFX_return_t(void)' [-Werror=enum-int-mismatch]
  392 | IFX_return_t ifx_mps_fastbuf_init (IFX_void_t)
      |              ^~~~~~~~~~~~~~~~~~~~
/builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0/src/mps/drv_mps_vmmc_common.c:120:13: note: previous declaration of 'ifx_mps_fastbuf_init' with type 'IFX_int32_t(void)' {aka 'int(void)'}
  120 | IFX_int32_t ifx_mps_fastbuf_init (IFX_void_t);
      |             ^~~~~~~~~~~~~~~~~~~~
/builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0/src/mps/drv_mps_vmmc_common.c:420:14: error: conflicting types for 'ifx_mps_fastbuf_close' due to enum/integer mismatch; have 'IFX_return_t(void)' [-Werror=enum-int-mismatch]
  420 | IFX_return_t ifx_mps_fastbuf_close (IFX_void_t)
      |              ^~~~~~~~~~~~~~~~~~~~~
/builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0/src/mps/drv_mps_vmmc_common.c:121:13: note: previous declaration of 'ifx_mps_fastbuf_close' with type 'IFX_int32_t(void)' {aka 'int(void)'}
  121 | IFX_int32_t ifx_mps_fastbuf_close (IFX_void_t);
      |             ^~~~~~~~~~~~~~~~~~~~~
[...]
make[7]: *** [scripts/Makefile.build:289: /builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0/src/mps/drv_mps_vmmc_common.o] Error 1
/builder/shared-workdir/build/staging_dir/host/bin/gcc -pthread -shared -L/builder/shared-workdir/build/staging_dir/host/lib -L/builder/shared-workdir/build/staging_dir/hostpkg/lib -L/builder/shared-workdir/build/staging_dir/target-mips_24kc_musl/host/lib -Wl,-rpath,/builder/shared-workdir/build/staging_dir/hostpkg/lib -Wl,--no-as-needed -lrt -L/builder/shared-workdir/build/staging_dir/host/lib -L/builder/shared-workdir/build/staging_dir/hostpkg/lib -L/builder/shared-workdir/build/staging_dir/target-mips_24kc_musl/host/lib -Wl,-rpath,/builder/shared-workdir/build/staging_dir/hostpkg/lib -Wl,--no-as-needed -lrt -fno-semantic-interposition -fprofile-generate -L/builder/shared-workdir/build/staging_dir/host/lib -L/builder/shared-workdir/build/staging_dir/hostpkg/lib -L/builder/shared-workdir/build/staging_dir/target-mips_24kc_musl/host/lib -Wl,-rpath,/builder/shared-workdir/build/staging_dir/hostpkg/lib -Wl,--no-as-needed -lrt -O2 -I/builder/shared-workdir/build/staging_dir/host/include -I/builder/shared-workdir/build/staging_dir/hostpkg/include -I/builder/shared-workdir/build/staging_dir/target-mips_24kc_musl/host/include -I/builder/shared-workdir/build/staging_dir/host/include/e2fsprogs -I/builder/shared-workdir/build/staging_dir/host/include -I/builder/shared-workdir/build/staging_dir/hostpkg/include -I/builder/shared-workdir/build/staging_dir/target-mips_24kc_musl/host/include -I/builder/shared-workdir/build/staging_dir/host/include/e2fsprogs build/temp.linux-x86_64-3.11/builder/shared-workdir/build/build_dir/hostpkg/Python-3.11.7/Modules/_opcode.o -L/builder/shared-workdir/build/staging_dir/host/lib -L/builder/shared-workdir/build/staging_dir/hostpkg/lib -o build/lib.linux-x86_64-3.11/_opcode.cpython-311-x86_64-linux-gnu.so
make[6]: *** [Makefile:1908: /builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0/src] Error 2
make[6]: Leaving directory '/builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/linux-5.15.153'
make[5]: *** [Makefile:1359: drv_vmmc.ko] Error 2
make[5]: Leaving directory '/builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0/src'
make[4]: *** [Makefile:440: all-recursive] Error 1
make[4]: Leaving directory '/builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0'
make[3]: *** [Makefile:360: all] Error 2
make[3]: Leaving directory '/builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0'
make[2]: *** [Makefile:187: /builder/shared-workdir/build/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/drv_vmmc-1.9.0/.built] Error 2
make[2]: Leaving directory '/builder/shared-workdir/build/package/kernel/lantiq/ltq-vmmc'
time: package/kernel/lantiq/ltq-vmmc/compile#27.76#3.73#34.73
    ERROR: package/kernel/lantiq/ltq-vmmc failed to build.
```
Ironically, enabling `CONFIG_KERNEL_WERROR` locally also didn't result in being able to reproduce the error -- I first had to do a `make target/linux/clean` to make that option have any effect on out-of-tree package builds...

So while I understand the need for feedback, setting this option **only** on buildbots is problematic. If literally just experienced the problematic aspects of it myself. Buildbots are mostly meant to provide us with ready binary snapshot builds, plus some logs in case something goes wrong. So who ever was bumping gcc supposedly also built all targets once before doing so, maybe even had github CI doing it for them. And the reason they didn't notice any error was supposedly because they didn't have `CONFIG_KERNEL_WERROR` enabled. Or they didn't care to fix a legacy target-specific out-of-tree package (which is also fine in that case).

Anyway, as you were calling it a trade-off, what would we loose if `CONFIG_KERNEL_WERROR` would always be enabled by default? Developers still always have to fix those warnings-becoming-errors (and that's fine imho), just now we also need to keep track of the default gcc version in another Kconfig file or otherwise manually enable `CONFIG_KERNEL_WERROR`, then `make target/linux/clean` and then see the error which would anyway occur on the buildbots.

A "trade" somehow implies (at least to me) that someone would gain something in exchange for having to give something. If you only lose and don't gain anything in exchange it ain't a good trade, right?

So, on the gaining side I understand what we want is feedback from the compiler about potential problems or bad code quality in (downstream) kernel patches or out-of-tree packages. That's good, and I agree that we do want that.

On the giving side, no matter what, given that `CONFIG_KERNEL_WERROR` is enabled on the buildbots, it means that we have to fix those errors by the end of the day, because otherwise it means the package won't be built.

So from OpenWrt developers' point of view, both the gaining and the giving side are always the same, no batter what the default setting on non-buildbot-builds is. Maybe not having it enabled by default is a bit worse (imho) because I had to skim through 26 MiB of buildbot pkgbuild stdio to get an idea what's happening and then found out the hard way that `CONFIG_KERNEL_WERROR` has a conditional default depending on whether building on buildbots or not and/or someone had forgotten to bump the version of GCC which would also enable it by default -- in my case that all clearly adds to the "giving" side while not getting anything in exchange on the "gaining" side. Imho that's exactly the "bad developer experience" you were mentioning, and yes, that divergence of the default is what made it worse.

Trying to understand your argument, maybe I have to understand how you define the "user" whose experience we care about. Is it a downstream developer who compiles OpenWrt from source and wants the build to succeed without being bothered by `-Werror` being set for kernel-related things, and without having to manually disable `CONFIG_KERNEL_WERROR` (they can always disable it in their `.config`, e.g. using menuconfig, obviously)?

Or is the "user" someone who wants to install a buildbot generated binary package on their hardware and not even mess around with sourcecode or build anything at all? Those "users" currently don't have a choice and need to rely on what buildbots generate.

The latter are more like courageous end-users messing with snapshot builds. The former are more like downstream developers. Imho both are "users" and should have a good experience, or at least we shouldn't deliberately make it worse without a good reason.

I admit I may not have understood your point fully, and I believe I may still lack information why you consider the current way as something different from "Having it (=KERNEL_WERROR) fully enabled" plus additional annoyance.
Where is that "trade off", as in: What are we giving (ok that's quite clear) and what are we getting in exchange which we couldn't also have (cheaper) when having it always enabled?

Do you understand my point at least...?